src/community/params-extractor/src/main/java/org/geoserver/params/extractor/Filter.java

```diff
@@ -7,58 +7,100 @@
+        // flags for this scenario
+        USE_AS_SERVLET_FILTER = true;
+        servletInstance = true;
+    }
+
     public Filter(GeoServerDataDirectory dataDirectory) {
-        Resource resource = dataDirectory.get(RulesDao.getRulesPath());
-        rules = RulesDao.getRules(resource.in());
-        resource.addListener(notify -> rules = RulesDao.getRules(resource.in()));
+        servletInstance = false;
+        initRules(dataDirectory);
+    }
+
+    private void initRules(GeoServerDataDirectory dataDirectory) {
+        if (dataDirectory != null) {
+            Resource resource = dataDirectory.get(RulesDao.getRulesPath());
+            rules = RulesDao.getRules(resource.in());
+            resource.addListener(notify -> rules = RulesDao.getRules(resource.in()));
+        }
     }
 
     @Override
     public int getPriority() {
         return ExtensionPriority.HIGHEST;
     }
 
+    /**
+     * This method is called only when the Filter is used as a standard web container Filter. When
+     * this happens the related instance will be used instead of the one initialized by Spring.
+     */
     @Override
-    public void init(FilterConfig filterConfig) throws ServletException {}
+    public void init(FilterConfig filterConfig) throws ServletException {
+        GeoServerDataDirectory dataDirectory =
+                GeoServerExtensions.bean(GeoServerDataDirectory.class);
+
+        initRules(dataDirectory);
+    }
 
     @Override
     public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
             throws IOException, ServletException {
-        HttpServletRequest httpServletRequest = (HttpServletRequest) request;
-        if (httpServletRequest.getRequestURI().contains("web/wicket")
-                || httpServletRequest.getRequestURI().contains("geoserver/web")) {
-            chain.doFilter(request, response);
-            return;
-        }
-        UrlTransform urlTransform =
-                new UrlTransform(
-                        httpServletRequest.getRequestURI(), httpServletRequest.getParameterMap());
-        String originalRequest = urlTransform.toString();
-        rules.forEach(rule -> rule.apply(urlTransform));
-        if (urlTransform.haveChanged()) {
-            Utils.info(
-                    LOGGER,
-                    "Request '%s' transformed to '%s'.",
-                    originalRequest,
-                    urlTransform.toString());
-            chain.doFilter(new RequestWrapper(urlTransform, httpServletRequest), response);
-        } else {
-            chain.doFilter(request, response);
+        if (isEnabled()) {
+            HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+            if (httpServletRequest.getRequestURI().contains("web/wicket")
+                    || httpServletRequest.getRequestURI().contains("geoserver/web")) {
+                chain.doFilter(request, response);
```

Comment:
Can we have just one call to doFilter? This if could have an else.

Reply:
done